[feat] 알람 등록 로직을 개발한다

oOccasio opened the pull request:
## #️⃣ 연관된 이슈
- close : #22 

## 📝 작업 내용
<!-- 이번 PR에서 작업한 내용을 설명해 주세요 -->
- 스케줄 키를 원래는 token 기준으로 생성했는데 동일 task에 token 별로 여러개 스케줄링이 필요없을 것 같아서 task 당 한개씩 하려고 task 기준으로 생성하도록 바꿨습니다
- 이렇게 되면 배치를 적용하기 어색해지기 때문에 현재는 배치적용은 뺐습니다
- 에러 로직은 저번과 같은 기준으로 삭제하는 로직 추가했습니다
- public 제어자 갖고 있는 메서드 기준으로 스케줄, 캔슬 호출하시면 될 것 같습니다

## 📸 스크린샷 (선택)
<!-- 실행 결과를 첨부해 주세요 -->

## 📢 참고 사항
- 현재 지금 aiplan -> plan -> user 를 갖고 있는데, 이중 LAZY이어서 한번 쿼리를 날려서 쿼리가 2번 이상 나오는지 확인하고 userId를 받는것을 고려하고 있습니다

고도화로 생각하고 있는게 
- 단일서버여도 장애시 복구와 확장성 측면에서, Event로 재스케줄링 하는 것보단 Redis로 관리하는 것이 좋다고 생각해서 그에 관한 로직 추가할것 같습니다
- 알람 보냈을 경우 실패한 경우 재시도로직 추가할것 같습니다
- 추가로 FCM 여러 서버 운영할 경우 Redis 분산락 부분도 건드리고 싶지만 다중서버 띄우거나 조금 오버엔지니어링인 감이 있을것 같아서 학습만 해볼 계획입니다

oOccasio committed:
feat(temp): #22 Notification Service

oOccasio committed:
feat(temp): #22 Notification Service

# Conflicts:
#	src/main/java/com/miruni/backend/domain/fcm/service/NotificationService.java

oOccasio committed:
Merge branch 'feat/#22' of https://github.com/UMC-8th-MIRUNI/MIRUNI_Backend_New into feat/#22

oOccasio committed:
feat: #22 NotificationService - Send, Execute, Schedule

oOccasio committed:
feat: #22 Refine Notification Scheduling And Token Handling
- add cancel notification logic
- delete invalid FCM tokens
- refine schedule key from token-based to task-based

oOccasio committed:
refactor: #22 Unused Import Delete

jjin70 commented on src/main/java/com/miruni/backend/domain/fcm/service/NotificationService.java:
```diff
@@ -1,21 +1,315 @@
+
+    //토큰 활성화 여부 체크
+    private boolean canReceiveAlarm(AlarmType alarmType, String token) {
+        FcmToken fcmToken = fcmTokenQueryService.getTokenByToken(token);
```
executeNotification에서 호출되느데 FcmToken 조회를 두 번 하게 되는 것 같습니다

oOccasio replied:
앗 그렇네요 반영했습니다

jjin70 commented on src/main/java/com/miruni/backend/domain/fcm/service/NotificationService.java:
```diff
@@ -1,21 +1,315 @@
+        NAG
+    }
+
+    private record NotificationContent(
+            String title,
+            String body
+    ){}
```
따로 분리 안 한 이유가 있나여??

oOccasio replied:
클래스 분리라면 여기서 밖에 쓰이지않을 것 같아서 분리하지 않았고 createContent를 다른 메서드로 2개만드는 것보다 하나로 만드는 것이 나을것 같아서 분리하지 않았습니다

jjin70 reviewed (approved):
수고하셨습니다!!

oOccasio committed:
refactor: #22 Duplicate Query Refactoring

oOccasio reviewed (commented):
두번 조회되는 부분 반영했습니다

DOHOON0127 committed:
rebase : notification rebase

# Conflicts:
#	src/main/java/com/miruni/backend/domain/fcm/service/NotificationService.java

DOHOON0127 committed:
[hotfix] mail관련 환경변수 추가

DOHOON0127 committed:
[hotfix] mail관련 시크릿 도커 컴포즈에 추가

DOHOON0127 committed:
[hotfix] mail 환경변수 인식 못하는 문제 수정

DOHOON0127 committed:
[hotfix] 테스트를 위한 403 허용

Suhyeon7 reviewed (approved):
수고하셨습니다!

oOccasio committed:
Merge branch 'dev' into feat/#22
